JBPM-5448 & JBPM-5811 & JBPM-5501: Improved evaluation of the next available position for newly added Graph elements.

alepintus opened the pull request:
Hi @romartin ,
 
this is the commit for the following tickets:

[Stunner - New elements added to business process have alignment to the top border 
](https://issues.jboss.org/browse/JBPM-5448
)
[Stunner - possibility to add several nodes after gateway without node moving
](https://issues.jboss.org/browse/JBPM-5811
)
[Stunner - Append new element doesn't work for elements inside Swimlanes
](https://issues.jboss.org/browse/JBPM-5501
)

Can you please review this PR?

Thanks.

hasys commented:
Jenkins execute full downstream build

romartin commented:
ok to test

romartin commented:
@hasys I think you should review this PR too, I will do it as well, let's keep in touch! :) Thanks!

romartin commented:
Hey @alepintus ,

Doing a try to this I found some issues:

- Expection popups - added some screenshots where trying to add a new task, by clicking on the selected shape's contextual menu button, it results in an error popup. For example as you can see when doing this from the embedded subprocess' shape, or when moving the gateway on canvas bottom right and clicking on new task.

- See screenshot about browser's debug console -  the `LayoutBoundExceededException` should result in some nice i18n message, rather than an error popup :)

![canvas-bottom-right](https://user-images.githubusercontent.com/4602417/27265142-67fade5a-5490-11e7-80b0-7bc74cc4a839.png)

![click-on-new-task-bug](https://user-images.githubusercontent.com/4602417/27265144-6fa5fdb0-5490-11e7-91b9-403448582939.png)

![some-debugs](https://user-images.githubusercontent.com/4602417/27265146-72a8b61a-5490-11e7-8dc9-186db52d9eb4.png)

Thanks!

hasys commented on kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-client-common/src/main/java/org/kie/workbench/common/stunner/core/client/canvas/util/CanvasLayoutUtils.java:
```diff
@@ -152,40 +177,80 @@ public static boolean isCanvasRoot(final Diagram diagram,
     @SuppressWarnings("unchecked")
     public double[] getNext(final CanvasHandler canvasHandler,
-                            final Node<View<?>, Edge> root) {
+                            final Node<View<?>, Edge> root,
```
If I correct there are no tests for this big bunch of tests at all. @alepintus can I ask you to write some?

romartin commented:
great, thanks for the review @hasys ! please @alepintus let us know once these comments are fixed so we can review again. Thanks!

alepintus commented:
Hi @romartin, @hasys

PR updated with buxfixes and with the new test.

Can you please review?

Thanks.

hasys commented:
Jenkins execute full downstream build

hasys commented:
Jenkins execute full downstream build

hasys commented on kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-client-common/src/test/java/org/kie/workbench/common/stunner/core/client/canvas/util/CanvasLayoutUtilsTest.java:
```diff
@@ -0,0 +1,252 @@
+    }
+
+    @Test
+    public void getNextFromRoot() {
```
Thank you for this test. Just one thing: can I ask you to run IDE Formatter over it (Ctrl+Alt+L) to avoid useless reformattings in the future?

P.S. Also, you can enable checkbox to make code reformat and Optimize imports automatically on commit. Click Ctrl+K and set check boxes as on picture.
![image](https://user-images.githubusercontent.com/1477262/27477090-696f806e-580b-11e7-8f66-e0c93e4bd54f.png)

hasys commented on kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-client-common/src/test/java/org/kie/workbench/common/stunner/core/client/canvas/util/CanvasLayoutUtilsTest.java:
```diff
@@ -0,0 +1,252 @@
+@RunWith(MockitoJUnitRunner.class)
+public class CanvasLayoutUtilsTest {
+
+    final double newNodeWidth = 100;
```
Every field in test class can be private to keep clean API.

hasys commented on kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-client-common/src/main/java/org/kie/workbench/common/stunner/core/client/canvas/util/CanvasLayoutUtils.java:
```diff
@@ -200,29 +246,31 @@ public static boolean isCanvasRoot(final Diagram diagram,
             }
         }
         final Bounds bounds = getGraphBounds(canvasHandler);
         final Bounds.Bound lr = bounds.getLowerRight();
```
After changes `lr` variable is not used any more. Is it expected and if so can you remove it, please?

alepintus commented:
Hi @hasys , @romartin ,
PR updated with all the fixes.

Can you please review?

Thanks.

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-client-common/src/main/java/org/kie/workbench/common/stunner/core/client/canvas/util/CanvasLayoutUtils.java:
```diff
@@ -93,6 +73,21 @@ public static boolean isCanvasRoot(final Diagram diagram,
+    public double[] getNext(final CanvasHandler canvasHandler,
+                            final double width,
+                            final double height,
+                            final double offsetX,
```
Please, instead of using two arguments (`offsetX`, `offsetY`), can you replace (where possible in this class) by a single `org.kie.workbench.common.stunner.core.graph.content.view.Point2D` argument, such as `Point2D offset. Reducing this way the number of arguments and the api for `Point2D` is better...

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-client-common/src/main/java/org/kie/workbench/common/stunner/core/client/canvas/util/CanvasLayoutUtils.java:
```diff
@@ -104,6 +99,8 @@ public static boolean isCanvasRoot(final Diagram diagram,
                            root,
                            width,
                            height,
+                           offsetX,
```
Replace by `Point2D`

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-client-common/src/main/java/org/kie/workbench/common/stunner/core/client/canvas/util/CanvasLayoutUtils.java:
```diff
@@ -114,12 +111,12 @@ public static boolean isCanvasRoot(final Diagram diagram,
                            nodeList,
-                           width,
                            height,
+                           offsetX,
```
Replace by `Point2D`

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-client-common/src/main/java/org/kie/workbench/common/stunner/core/client/canvas/util/CanvasLayoutUtils.java:
```diff
@@ -128,6 +125,8 @@ public static boolean isCanvasRoot(final Diagram diagram,
     public double[] getNext(final CanvasHandler canvasHandler,
                             final double w,
                             final double h,
+                            final double offsetX,
```
Replace by `Point2D`

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-client-common/src/main/java/org/kie/workbench/common/stunner/core/client/canvas/util/CanvasLayoutUtils.java:
```diff
@@ -139,6 +138,8 @@ public static boolean isCanvasRoot(final Diagram diagram,
                            root,
                            w,
                            h,
+                           offsetX,
```
Replace by `Point2D`

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-client-common/src/main/java/org/kie/workbench/common/stunner/core/client/canvas/util/CanvasLayoutUtils.java:
```diff
@@ -150,42 +151,79 @@ public static boolean isCanvasRoot(final Diagram diagram,
                            nodeList,
-                           w,
                            h,
+                           offsetX,
```
Replace by `Point2D`

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-client-common/src/main/java/org/kie/workbench/common/stunner/core/client/canvas/util/CanvasLayoutUtils.java:
```diff
@@ -150,42 +151,79 @@ public static boolean isCanvasRoot(final Diagram diagram,
+        final double[] newNodeSize = GraphUtils.getNodeSize(newNode.getContent());
+        final Point2D newNodePosition = GraphUtils.getPosition(newNode.getContent());
+        final double[][] offset = new double[1][1];
+        final double[] offsetX = {0};
```
Replace by `Point2D`

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-client-common/src/main/java/org/kie/workbench/common/stunner/core/client/canvas/util/CanvasLayoutUtils.java:
```diff
@@ -150,42 +151,79 @@ public static boolean isCanvasRoot(final Diagram diagram,
+        final double[][] offset = new double[1][1];
+        final double[] offsetX = {0};
+        final double[] offsetY = {0};
+        final double[] parentOffsetX = {0};
```
Replace by `Point2D`

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-client-common/src/main/java/org/kie/workbench/common/stunner/core/client/canvas/util/CanvasLayoutUtils.java:
```diff
@@ -198,35 +236,30 @@ public static boolean isCanvasRoot(final Diagram diagram,
                              final List<Node<View<?>, Edge>> nodes,
                              final double width,
-                             final double height,
+                             final double offsetX,
```
Replace by `Point2D`

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-client-common/src/main/java/org/kie/workbench/common/stunner/core/client/canvas/util/CanvasLayoutUtils.java:
```diff
@@ -236,48 +269,29 @@ public static boolean isCanvasRoot(final Diagram diagram,
-                           final double uy) {
+    private double[] getNextPositionWithOffset(final double x,
+                                               final double y,
+                                               final double offsetX,
```
Replace by `Point2D`

hasys commented on kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-client-common/src/test/java/org/kie/workbench/common/stunner/core/client/canvas/util/CanvasLayoutUtilsTest.java:
```diff
@@ -0,0 +1,278 @@
+@RunWith(MockitoJUnitRunner.class)
+public class CanvasLayoutUtilsTest {
+
+    private final double newNodeWidth = 100;
```
It is better to keep convention and use uppercase letters in constants. See [1] `Constants` section for more details.

[1] https://docs.oracle.com/javase/tutorial/java/javaOO/classvars.html

hasys commented on kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-client-common/src/test/java/org/kie/workbench/common/stunner/core/client/canvas/util/CanvasLayoutUtilsTest.java:
```diff
@@ -0,0 +1,278 @@
+    private final double offsetX = 100;
+    private final double offsetY = 100;
+    @Mock
+    Diagram diagram;
```
All mocks and any other fields in Test class, in general, have to be `private`.

alepintus commented:
Hi @hasys , @romartin ,
Refactored PR using Point2D instead of doubles and fixed issues regarding code convention.

Can you please review?

Thanks.

romartin commented:
Jenkins please retest this
